쥬스 메이커 [STEP 3] sookoong, bard

Jeon-Minsu opened the pull request:
안녕하세요 제이크!  @jryoun1 
벌써 마지막 STEP이네요🥲  
이번 STEP도 잘부탁드리겠습니다

## 고민한 점🤔
### 1. `BarButtonItem, ButtonItem`
- `BarButtonItem`을 사용하여 과일 재고를 관리해주는 뷰컨트롤러에서 쥬스를 만드는 뷰 컨트롤러로 돌아오는 기능을 수행하는 버튼을 만들어 줘야 했었는데, `Button`레이블을 생성하여 오른쪽 상단에 넣어줬더니 마치 `Right BarButton Item`으로 귀속되는 것처럼 들어가 `Button Item`을 `BarButton Item`으로 착각해서 코드를 작성하였었습니다.  
![ButtonItem](https://user-images.githubusercontent.com/92622931/167352740-b64a869d-8b22-4fd1-98b7-1902ccbc00ab.gif)  
- 이렇게 만들어진 버튼은 `sender`를 `UIButton`으로 설정해 주었어야 했는데, `Button`이 들어간 영역이 `BarButtonItem`영역으로 들어가는 줄 착각하고 `sender`를 `UIBarButtonItem`으로 적어 왜 안되는지 한참을 고민했었습니다😵‍💫
- `document outline`이나 `Attirbutes Inspector`를 조금만 더 상세히 봤다면 고민할 일이 없었을 텐데 저희가 쓴 코드가 틀린 부분이 없다고 생각해 한참을 고민을 하다가 나중이 되서야 이유를 알게되었습니다


### 2. `UI Object` 
- `UI Object`들을 스토리보드로 만들 경우 개인이 할 때는 편하게 만들 수 있지만, 
- 협업할 때에는 하나하나 찾아봐야 하기 때문에 코드로 작성하는 것이 협업하는데 편하다는 장점이 있습니다.
- 현 프로젝트에서는 둘이서 라이브 코딩으로 협업하기에 스토리보드를 사용하여 만들어보았습니다.

### 3. `뷰컨트롤러에서 다른 뷰 컨트롤러에서 데이터를 전달해주는 법`
- 쥬스를 만들어주는 뷰 컨트롤러에서는 `property type`를 사용해 데이터를 전달해주었고, 반면 과일 재고를 관리하는 뷰 컨트롤러에서는 `notification`을 사용해 데이터를 전달해주었습니다.
- 뷰 컨트롤러 간의 데이터를 전달하는 방법을 6가지 방법이 있는것으로 알고 있는데  
	- `property에 직접 접근` 
		- 간단한 방법이지만, `push`, `present`방식으로 화면을 전환하는 경우에만 데이터가 정상적으로 넘어감
	- `property와 함수를 이용`
		- 서로 함수에 의존해서 데이터를 전달하는 방식이기에 유지보수가 어려움
	- `segue` 이용
		- 컨트롤러에 직접적으로 의존하고 있는 형태
		- 뷰컨트롤러에 강하게 결합되어있어 수정할 때 어려움
	- `delegate` 이용
		- 위의 방법처럼 강한 결합한 상태를 보완하기 위한 방법
		- 뷰 컨트롤러간의 의존을 하지않고 떨어져있는 구조
	- `closure` 이용
		- `delegate`보다 간결하며 프로토콜 및 메소드 없이 지역 스코프 내에서 바로 처리할 수 있음
	- `notificationCenter`
		- 간단한 데이터를 넘기거나, 어떠한 변화에 대응하는 작업을 할 때 사용하는 것이 적절
		- 모든 데이터를 `notificationCenter`로 넘긴다면 여러 변화를 `observe`해야 하므로 상당히 비효율적인 방식이 될수도 있음

- 이 6가지 방식 중 간단한 데이터만 넘기고 현재 뷰 컨트롤러를 `present`형식으로 넘겨주고 있어서 `propery`를 이용해 데이터를 넘겨주어보았고, `notificationCenter` 같은 경우는 약 1주일 가량 씨름을 해와서 사용한 것이 없지않아 있습니다만(*아래에 질문으로 따로 남겼습니다*), 간단한 데이터를 넘기고 변화에 대응하는 작업을 할 때 사용했기에 현 프로젝트에서 사용해보았습니다.
	



---

## 질문하고 싶은 것🙋


### 1. `Notification` 관련 질문

>- 화면 이동 방식: ViewController 1 -> ViewController 2 -> ViewController 1 
>- 데이터 전달 구조 : ViewController2 -> ViewController1

- VC2 에서 Notification을 post,  
VC1 에서 addObserver를 구현하려고 한다면,

- VC1에서 NotificationCenter에 Observer로 등록을 해놓았기 때문에   
VC2에서 NotificationCenter로 알림을 Post하면,   
미리 VC1에서 등록한 옵저버에서 해당 알림을 수신하여, selector에 등록한 기능을 수행할 수 있습니다.

- 반면에, 

>- 화면 이동 방식: ViewController 1 -> ViewController 2 -> ViewController 1 
>- 데이터 전달 구조 : ViewController1 -> ViewController2


- VC1 에서 Notification을 post,   
VC2 에서 addObserver를 구현하려고 한다면,

- VC1에서 NotificationCenter로 알림을 Post하였으나, 메모리에 옵저버가 등록되지 않았기 때문에 해당 알림을 수신을 받지 못합니다.   
VC2에서 뒤늦게 NotificationCenter에 Observer로 등록을 해보았지만, 이미 VC1에서 보낸 post는 만료가 되어 VC2의 Observer는 아무런 Notification을 수신하지 못하는 상태입니다.

- 즉, 문제를 정리하자면, Observer가 메모리에 올라가기 이전에 VC1에서 post를 하였기에 observer가 알림을 인식하지 못하는 상태에 직면한 것입니다.

- 이를 해결하기 위해서는 VC1에서 알림을 post하기 이전에 observer를 메모리에 먼저 올린다면 VC2에서도 알림을 수신할 수 있을 것입니다. 하지만, 어떻게 VC1에서 알림을 post하기 전에 VC2의 Observer를 메모리에 먼저 올릴 수 있는지 방법을 아무리 궁구해보아도 떠오르지 않았습니다.

- 수일동안 여러 차례 고민하여 방법을 궁구하던 중 학습활동 시간을 통해 View Life Cycle의 State Method가 각각 언제 호출되는지 탐구하던 중 다음의 생각이 떠올랐습니다.

- VC1에서 VC2로 기본 모달방식으로 화면전환을 할 때 각각의 State Method는 다음과 같이 호출됩니다.

[모달 방식 - automatic]
```swift
1st viewDidLoad
1st viewWillAppear
1st viewDidAppear
======================
2nd viewDidLoad
2nd viewWillAppear
2nd viewDidAppear
```

- VC1에서 VC2로 풀스크린 모달방식으로 화면전환을 할 때 각각의 State Method는 다음과 같이 호출됩니다.

[모달 방식 - fullScreen]
```swift
1st viewDidLoad
1st viewWillAppear
1st viewDidAppear
======================
2nd viewDidLoad
1st viewWillDisappear
2nd viewWillAppear
2nd viewDidAppear
1st viewDidDisappear
```

- 그렇다면, VC1이 viewWillDisappear나 viewDidDisappear에서 Notification을 Post하면
이미 VC2에서 '2nd viewDidLoad'로 옵저버를 등록하면 알림을 수신받을 수 있지 않을까 하는 가설을 생각하게 됩니다.
그리고 이를 구현해본 결과 VC2의 observer에 등록된 selector 함수가 올바르게 실행됨을 확인할 수 있었습니다!!

- 이에 제이크에게 해당 방법이 앞서 끊임없이 갈구하였던 메모리에 미리 observer를 올린 알맞은 방법에 해당하는지 궁금합니다. 추가적으로 혹시나 다른 방법으로는 어떻게 미리 Observer를 메모리에 올릴 수 있는지 또한 궁금하여 질문드립니다!

---
### 2 - 1 `Modal` 내 `Navigation` 구현 중 `Navigation Bar` 미표시

- 쥬스 메이커 프로젝트에서 과일 주문 VC(=`JuiceOrderViewController`)와 재고 수정   
- VC(=`FruitStockViewController`)를 생성하였는데, 처음에는 전자에서 후자로 VC가 이동하는 방식을 Navigation으로 정의하였으나, 재고 수정은 과일 주문이라는 흐름을 끊고 별도의 행동을 취하는 것으로 생각이 바뀌게 되어 `Modal` 방식으로 이를 다음과 같이 수정하였습니다. 

```swift
private func goToFruitStockViewController() {
    guard let controller = self.storyboard?.instantiateViewController(withIdentifier: "FruitStockViewController") as? FruitStockViewController else { return }
    
    present(controller, animated: true)
}
```

- 그러다보니 이전 방식에서 `Navigation Bar`의 `Title` 과 뒤로가기 버튼이 자동적으로 생성되었는데, `Modal` 방식에는 이를 새로 생성할 필요가 있었습니다. 하지만, 아무리 스토리보드 내 재고 수정 `VC`에 `Title`과 `Button`을 표시하여도 `Simulator` 작동시에는 표시되지 않아 이를 해결할 필요가 있었습니다.

- 결론을 앞서 밝히면, 바로 `재고 수정 VC`로 화면을 전환하지 않고, 원하는 `VC`를 `Navigation Controller` 내에 `root VC`로 설정한채 `navigation controller`로 경로(`identifier`)를 지정하여 화면을 전환하면 `Navigation Bar`의 누락을 피할 수 있었습니다.


- 다시 말해, `Navigation Controller`로 이동하지 않고 바로 `VC`로 `Identifier`을 설정하면 네비게이션 바가 생략된 채 화면이 전환되지만, 이를 `Navigation Controller`으로 경로를 지정함과 동시에 `root VC`로 원하는 `VC`를 넣었다면, 네비게이션의 제목과 버튼들이 생략되지 않고 화면 전환에 성공할 수 있었습니다.
>

```swift
private func goToFruitStockViewController() {
    guard let controller = self.storyboard?.instantiateViewController(withIdentifier: "ModalViewController") else { return }
    
    present(controller, animated: true)
}
```

- 이에, 스토리보드를 보면 `FruitStockViewController`에 `navigation title`과 `item`들이 정상적으로 표시되는데도 불구하고 네비게이션 컨트롤러가 아닌 `VC`로 화면 전환을 바로하면 왜 네비게이션 바가 노출되지 않는지 이유가 궁금합니다.


### 2 - 2 `Modal`을 통한 의존성 주입 방법
- 추가로 이는 다음의 문제와도 관련이 있습니다.

- Step3 구현 중 `JuiceOrderViewController` 내의 과일 `label`의 `text`를 `FruitStockViewController`로 전달하기 위한 방법을 생각하였습니다. 처음에는 `Navigation` 방식에서 의존성 주입을 통해 아래와 같이 식을 작성하면 `FruitStockViewController` 내 프로퍼티에 과일 `label`의 `text`를 할당함과 동시에 전달할 수 있었습니다.

```swift
@IBAction func editStock(_ sender: UIBarButtonItem) {
    guard let controller = self.storyboard?.instantiateViewController(withIdentifier: "FruitStockViewController") as? FruitStockViewController else { return }
    
    controller.text1 = strawberryLabel?.text ?? ""
    controller.text2 = bananaLabel?.text ?? ""
    controller.text3 = pineappleLabel?.text ?? ""
    controller.text4 = kiwiLabel?.text ?? ""
    controller.text5 = mangoLabel?.text ?? ""
    
    self.navigationController?.pushViewController(controller, animated: true)
   
}
```

- 하지만 모달 방식을 이용하여 네비게이션 기능을 이용하다 보니 위와 비슷한 형식으로 코드를 작성하였으나 데이터 전달은 이루어졌으나 네비게이션 바가 사라지고 말았습니다.

```swift
@IBAction func editStock(_ sender: UIBarButtonItem) {
    guard let controller = self.storyboard?.instantiateViewController(withIdentifier: "FruitStockViewController") as? FruitStockViewController else { return }
    
    controller.text1 = strawberryLabel?.text ?? ""
    controller.text2 = bananaLabel?.text ?? ""
    controller.text3 = pineappleLabel?.text ?? ""
    controller.text4 = kiwiLabel?.text ?? ""
    controller.text5 = mangoLabel?.text ?? ""
    
    present(controller, animated: true)
   
}
```
-   문제는 `바로 Navigation Controller`로 화면 전환을 이루지 않고, `FruitStockViewController`로 바로 화면 전환을 이루었기 때문이었습니다. 모달 방식을 이용하는 가운데 바로 `root viewController`로 데이터를 전달하면 프로퍼티를 통해 데이터 전달은 잘 이루어졌습니다. 하지만 네비게이션 바가 제대로 노출되지 않거나 기능하지 않는 경우가 발생하게 된 것이었습니다.

-  모달 방식을 이용할 때 바로 뷰컨트롤러로 화면 전환을 하면 네비게이션 바가 잘 작동하지 않는다는 사실은 알게 되었습니다. 그러면 경로를 네비게이션 컨트롤러로 설정한다고 문제가 해결될 것이라고 판단하하였지만 그렇지 않았습니다. 컴파일 오류가 발생하였습니다.
```swift
private func goToFruitStockViewController() {
    guard let controller = self.storyboard?.instantiateViewController(withIdentifier: "ModalViewController") else { return }
        
    controller.numberOfStrawberry = strawberryLabel.text ?? ""
    controller.numberOfBanana = bananaLabel.text ?? ""
    controller.numberOfPineapple = pineappleLabel.text ?? ""
    controller.numberOfKiwi = kiwiLabel.text ?? ""
    controller.numberOfMango = mangoLabel.text ?? ""

    present(controller, animated: true)
}
```
- 네비게이션 컨트롤러 모달을 하니 네비게이션 기능은 잘 작동하였으나 프로퍼티, 즉 데이터 전달이 이루어지지 않았습니다. 이는 `push`나 `present`의 경로가 `root VC`로 설정되어 있을때 인스턴스 프로퍼티가 전달될 수 있기 때문입니다. 그렇다면 두 방식의 이점만을 살리는 새로운 방법을 도출할 필요가 있었습니다.

- 위의 문제점을 해결할 한 가지 방법이 떠올랐습니다. 경로를 `Navigation Controller`로 설정하고 타입 프로퍼티를 통해 데이터를 전달한다면 네비게이션 바 기능의 이점과 프로퍼티 전달의 이점을 적극 활용할 수 있었습니다. 인스턴스 생성과 전혀 상관 없이 타입 프로퍼티를 최초 호출하면 메모리에 올라가 초기화가 이루어져 그 이후에는 어디서든 해당 프로퍼티 공유가 되기 때문에 이러한 방법을 사용할 수 있었습니다.

```swift
private func goToFruitStockViewController() {
    guard let controller = self.storyboard?.instantiateViewController(withIdentifier: "ModalViewController") else { return }
        
    dispatchFruitLabel()
    present(controller, animated: true)
}
    
private func dispatchFruitLabel() {
		FruitStockViewController.numberOfStrawberry = strawberryLabel.text ?? ""
    FruitStockViewController.numberOfBanana = bananaLabel.text ?? ""
    FruitStockViewController.numberOfPineapple = pineappleLabel.text ?? ""
    FruitStockViewController.numberOfKiwi = kiwiLabel.text ?? ""
    FruitStockViewController.numberOfMango = mangoLabel.text ?? ""
}
```

- 정리해보자면, 각각의 방식은 다음의 장단점을 가졌습니다. 


1) 모달로 바로 `root` 뷰컨으로 보내기  
장점: 프로퍼티가 잘 전달됨  
단점:  네비게이션 바 기능들이 잘 작동 X


2) 모달을 `navigation controller` 보내기  
장점: 네비게이션 바 기능 잘 작동  
단점: 프로퍼티 전달의 어려움  


3) 모달을 `navigation controller` 보내기 + 타입 프로퍼티 이용  
네비게이션 바 표현 가능, 프로퍼티 전달 가능  

- 이에 저희가 사용한 방법은 통상 이용하는 방법과 차이가 있지 않을까 염려가 생겨 다음의 방법을 사용하는 것이 가장 좋은 방법인지 궁금하여 질문드립니다.

---

### 3. `Navigation Bar Height`  
![Screen Shot 2022-05-09 at 4 21 21 PM](https://user-images.githubusercontent.com/92622931/167360107-3b02923c-1f01-45c0-845e-f87a725d6b60.png)  

- 프로젝트 기획서 내에는 `Navigation Bar`의 높이가 저희가 만든 `Navigation Bar`의 높이보다 길어보여서 `Navigation Bar Height`의 높이를 조절해보려고 하였는데, 저희가 찾은 정보로서는 `Navigation Bar`를 커스텀하는 방법 말고는 찾지를 못하였습니다.
- 소스코드에 들어가서 `Navigation Bar`에 해당하는 `Object ID`를 찾아 `Height`을 설정해보았는데 설정한대로 변하지를 않았습니다.
- `Navigation Bar`를 커스터마이징 하는것은 아직 저희 수준에서는 무리라 생각을 하였고, 이 부분은 건들지 않기로 결정하였습니다. 

- 혹시 커스터마이징 말고 다른 방법이 또 있나해서 여쭤봅니다! 
- 추가로 `stepper`같은 경우도 버전이 업그레이드가 되면서 아이콘이 변경되어 따로 변경할려면 이미지를 다운받아 집어넣어야 하는지 궁금합니다.

Jeon-Minsu committed:
feat: FruitStockViewController 생성 및 과일 label, stepper 추가

Jeon-Minsu committed:
feat: 화면전환 방식을 modal 방식으로 수정 및 FruitStockVIewController의 Navigation 바에 닫기 버튼 추가

bar-d committed:
feat: 스테퍼 버튼 FruitStockViewController에 연결

Jeon-Minsu committed:
feat: JuiceOrderViewController의 과일 label을 전달하여 FruitStockViewController의 과일 label 동기화

bar-d committed:
feat: FruitStoreViewController의 레이블과 각각의 stepperValue를 업데이트하는 함수 생성

Jeon-Minsu committed:
feat:  FruitStockViewController 내 타입 프로퍼티의 값 할당 및 post 기능 수행

bar-d committed:
feat:  접근제한 설정 및 observer 추가
- Notification post를 감지하는 observer 추가
- 과일의 재고를 업데이트할 수 있는 tryUpdateFruitStock함수 추가

Jeon-Minsu committed:
style: 접근 제한 설정, 네이밍 변경 및 함수 분리

bar-d committed:
style: 오토레이아웃 설정

Jeon-Minsu committed:
style: Navigation Bar 색상 변경, 주문 버튼 줄바꿈. 정렬 및 글자크기 변경

bar-d committed:
style: 글씨 크기 변경 및 네비게이션 백그라운드 색상 설정

bar-d committed:
refactor: 함수 은닉화 및 Notification center에 post해주는 함수 생성

Jeon-Minsu committed:
docs: README.md 파일 수정

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -6,30 +6,41 @@
+    @IBOutlet private weak var kiwiLabel: UILabel!
+    @IBOutlet private weak var mangoLabel: UILabel!
+    
+    @IBOutlet weak var orderButtonOfStrawberryBananaJuice: UIButton!
```
`strawberryBananaJuiceButton` 이어도 충분히 알 수 있을 것 같아요! 🙂

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -6,30 +6,41 @@
 
 import UIKit
 
+let center = NotificationCenter.default
```
NotificationCenter를 사용하셨군요 👏
클래스 밖에서 따로 선언하신 이유가 있을까요? 👀

Jeon-Minsu replied:
서로 다른 ViewController간 공통적으로 사용하기 위하여 다음과 같이 전역 상수로 선언하였습니다.

jryoun1 replied:
전역 상수를 사용하면 모든 곳에서 접근이 가능한데 전역 상수를 사용하는 것이 바람직할까요? 👀
물론 그럴 필요가 있는 객체라면 singleton 객체로 만들어서 사용할 수 있지만, 
그렇지 않다면 전역상수를 사용하는 것은 지양하는 것이 바람직합니다.

추가로 `NotificationCenter.default` 의 경우에는 싱글톤으로 구현되어 있습니다.

Jeon-Minsu replied:
전역 상수 삭제 완료하였습니다!

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -92,8 +102,68 @@ class JuiceOrderViewController: UIViewController {
+        try? juiceMaker.fruitStore.changeStock(fruit: .mango, amount: Int(FruitStockViewController.numberOfMango) ?? 0)
+    }
+    
+    private func breakLineOfButton() {
```
함수가 하는 역할을 보니 button의 titleLabel의 attribute를 설정해주는 함수인 것 같은데, 해당 함수명은 명확해보이지 않습니다!

bar-d replied:
`setUpAtttributesOfButton`으로 변경해주었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -92,8 +102,68 @@ class JuiceOrderViewController: UIViewController {
+    }
+    
+    private func breakLineOfButton() {
+        orderButtonOfStrawberryBananaJuice.titleLabel?.lineBreakMode = .byWordWrapping
```
버튼의 종류만 변경되지 나머지 설정해주는 부분들은 다 동일한 것 같은데 어떻게 하면 중복을 줄일 수 있을까요? 👀
hint: UIButton을 extension 해보면 어떨까요?

Jeon-Minsu replied:
UIButton에 extension을 추가하여 lineBreakMode, textAlignment, font를 원하는 조건으로 설정하여 코드 중복을 최소화하였습니다!

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -92,8 +102,68 @@ class JuiceOrderViewController: UIViewController {
+        showFruitStock()
+    }
+    
+    private func tryUpdateFruitStock() {
```
updateFruiteStock() 으로 충분할 것 같은데 try가 붙은 이유가 있을까요?

bar-d replied:
해당 함수내에서 `try`를 쓰고 바로 위에 있는 함수가 `updateFruitStock`이라는 함수가 따로 있어서 썼습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -92,8 +102,68 @@ class JuiceOrderViewController: UIViewController {
     }
 }
+
+extension Notification.Name {
```
Notification.Name도 사용하셨군요! 👏

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -6,30 +6,41 @@
-        kiwiLabel?.text = String(juiceMaker.fruitStore.fruitWarehouse[.kiwi] ?? 0)
-        mangoLabel?.text = String(juiceMaker.fruitStore.fruitWarehouse[.mango] ?? 0)
+
+    private func showFruitStock() {
```
이 함수는 stock을 (show)보여주는걸까요? 
보여주는건 alert를 보여주거나, action sheet 같은게 보여질때 보여진다고 하지 않을까요?? 👀

bar-d replied:
`updateFruitStockLabelText`로 함수명 변경해주었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -92,8 +102,68 @@ class JuiceOrderViewController: UIViewController {
+        present(controller, animated: true)
+    }
+    
+    private func dispatchFruitLabel() {
```
fruitelabel을 보내는건가요? 함수명이 명확하지 않은 것 같습니다.

bar-d replied:
전달해주는 게 더 명확한것 같아서 `sendFruitLabelText`로 변경해주었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -92,8 +102,68 @@ class JuiceOrderViewController: UIViewController {
+        present(controller, animated: true)
+    }
+    
+    private func dispatchFruitLabel() {
```
만약 과일의 개수가 극단적으로 100개로 늘어난다면 FruitStockViewController과 dispatchFruitLabel() 함수 내부에는 100줄의 코드가 추가될 것 같은데, 현재 이 방법이 좋은 방법일까요? 🤔

모든 과일들을 직접적으로 보내지 않고 관리하는 주체 하나를 주고 받아도 충분해 보입니다.

Jeon-Minsu replied:
데이터를 전달하는데 있어서 원래의 뷰컨트롤러 내에 있는 지엽적인 정보만을 보내는 데 주목을 하고 있었습니다. 하지만 제이크의 코멘트 덕분에 구조체, 클래스 등의 타입 인스턴스들도 데이터 전달이 가능함을 알 수 있었습니다. 이에 객체 자체를 넘겨주게되어 각각의 과일들에 대해서 `label`을 언급할 필요가 사라져 코드 중복을 피할 수 있었습니다!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -13,7 +13,7 @@ class FruitStore {
     ]
     
-    func checkInvalidAmount(_ amount: Int) throws {
+    private func checkInvalidAmount(_ amount: Int) throws {
```
외부에서 접근할 일이 없다면 항상 private을 붙여주는게 좋습니다! 👏

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -34,7 +34,7 @@ class FruitStore {
     }
     
-    func reduceStock(fruit: Fruit, amount: Int) throws {
+     func reduceStock(fruit: Fruit, amount: Int) throws {
```
스페이스는 의도한신 걸까요? 👀

bar-d replied:
수정했습니다 ㅎㅎ

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    @IBOutlet private weak var kiwiStockStepper: UIStepper!
+    @IBOutlet private weak var mangoStockStepper: UIStepper!
+    
+    static var numberOfStrawberry = ""
```
이 변수들은 static으로 선언되었는데 그 이유가 있을까요? 👀

Jeon-Minsu replied:
앞서 전체 PR 질문하고 싶은 2-2에서 일전의 방법에서는 인스턴스 프로퍼티로 제대로 전달이 이루어지지 않아, 모달 경로를 Navigation Controller로 설정하고 타입 프로퍼티를 통해 데이터를 전달한다면 네비게이션 바 기능의 이점과 프로퍼티 전달의 이점을 적극 활용할 수 있어 변수를 static으로 선언하였습니다. 타입 프로퍼티는 인스턴스 생성과 전혀 상관 없이 타입 프로퍼티를 최초 호출하면 메모리에 올라가 초기화가 이루어져 그 이후에는 어디서든 해당 프로퍼티 공유가 되기 때문에 이러한 방법을 사용할 수 있었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -92,8 +102,68 @@ class JuiceOrderViewController: UIViewController {
+        FruitStockViewController.numberOfMango = mangoLabel.text ?? ""
+    }
+    
+    private func observeStockChanges() {
```
이 함수는 내부에서는 NotficationCenter.default에 Observer를 달아주고 있네요.
Notification의 observer는 **등록한다**라고 많이 하는 것 같아요.
`register~` 라는 네이밍은 어떤가요?

bar-d replied:
`registerStockChanges`로 변경해주었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -92,8 +102,68 @@ class JuiceOrderViewController: UIViewController {
+        center.addObserver(self, selector: #selector(updateFruitStock), name: .checkFruitStock, object: nil)
+    }
+    
+    @objc private func updateFruitStock() {
```
private을 잘 사용해주셨네요 👍🏻

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -92,8 +102,68 @@ class JuiceOrderViewController: UIViewController {
 }
+
+extension Notification.Name {
+    static let checkFruitStock = Notification.Name("checkFruitStock")
```
해당 Notification은 check를 하는게 아니고 update를 해주고 있는 것 같습니다! 🙏🏻

bar-d replied:
`updateFruitStock`로 변경해주었습니다

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,95 @@
+        FruitStockViewController.numberOfMango = mangoLabel.text ?? ""
+    }
+    
+    private func postCheckFruitStockNotification() {
```
이 함수는 stepper 버튼이 눌릴때마다 매번 호출되어야하는 걸까요? 🤔

bar-d replied:
닫기 버튼을 눌릴때, 한번만 호출되면 된다고 생각합니다!  
이에 닫기 버튼 함수에 호출되도록 넣어주었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,95 @@
+        center.post(name: .checkFruitStock, object: nil)
+    }
+    
+    @IBAction private func strawberryStepper(_ sender: UIStepper) {
```
함수명은 명사보다는 동사에 가깝습니다!

bar-d replied:
함수명 앞에 `changeValueOf`를 넣어주었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,95 @@
+        postCheckFruitStockNotification()
+    }
+    
+    @IBAction private func closeCurrentView(_ sender: UIBarButtonItem) {
```
`closeButtonTouchedUpInside()` 
`closeButtonPressed()` 
와 같은 네이밍은 어떨까요? 

일반적인 함수라면 closeCurrentView()와 같은 네이밍도 괜찮지만, @IBAction이나 @objc 와 같은 함수들은 이벤트나 어떠한 행위에 의해서 발생하므로 주로 해당 행위를 나타내는 네이밍을 사용합니다.

bar-d replied:
`closeButtonTouchedUpInside`로 변경해주었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    @IBOutlet private weak var kiwiLabel: UILabel!
+    @IBOutlet private weak var mangoLabel: UILabel!
+    
+    @IBOutlet private weak var strawberryStockStepper: UIStepper!
```
네이밍 명확하고 좋네요 👍🏻

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -6,30 +6,41 @@
+    @IBOutlet private weak var kiwiLabel: UILabel!
+    @IBOutlet private weak var mangoLabel: UILabel!
+    
+    @IBOutlet weak var orderButtonOfStrawberryBananaJuice: UIButton!
```
이 버튼들은 외부에서 접근하는 경우가 있을까요? 👀

bar-d replied:
은닉화 하였습니다!

jryoun1 reviewed (changes requested):
@Jeon-Minsu @bar-d 
벌써 3주가 후다닥 지나갔네요🥲 
이번 step을 진행하면서 두 분이 정말 많은 고민을 해보시고 여러 가지를 시도해보신 걸 확인해 볼 수 있네요👍🏻💯
고생하셨습니다!

## 고민한 점
2. 벌써 두 개의 장단점 일부를 느끼셨네요 👍🏻
스토리보드로 사용했을 때의 장단점과 코드로 작성했을 때의 장단점에 대해서 앞으로 더 많이 느끼실 수 있을 겁니다! 

3. 가능한 방법들을 다 찾아보셨네요 👏

## 질문
1. 질문만봐도 많은 고민을 했다는 것이 느껴지네요 👍🏻 고생하셨습니다~
가설과 구현을 통해서 직접 확인도 해보시다니 멋있습니다 💯
다만 notification을 사용하는 경우에 대해서 차차 알아가시면 좋을 것 같습니다. 
notification에 대한 [공식문서](https://developer.apple.com/documentation/foundation/notificationcenter)를 확인해보면 

> A notification dispatch mechanism that enables the broadcast of information to registered observers.

이런 설명이 있습니다. 여기서 broadcast라는 것이 중요한 부분입니다. 
notification은 어떠한 이벤트가 발생했을 때 이를 알아야하는 여러 개의 객체들이 있을 때 이들 각각에게 알려주는 것은 코드의 양이 많아질 수 있으므로 한 번의 post로 인해 observer하고 있는 모든 곳에서 이를 알아차릴 수 있도록 1:다 에 적합한 방법입니다. 

그렇다면 이번 프로젝트와 같이 vc끼리 1:1로 데이터를 주고 받는 경우에는 어떠한 방법들을 사용할까요? 
위에서 말씀해주신 6가지 방법들 중 notificationCenter를 제외한 방법들을 사용할 수 있을 겁니다! 
물론 notification을 사용해도 됩니다. 그러나 지금 겪는 문제처럼 observer를 등록하는 시점과 post하는 시점이 순차적이지 않은 경우에 이를 수정하는 방법까지 고려하면서 notification을 사용했을 때 얻는 이점이 있을 것 같지는 않습니다. 

이외의 방법인 delegate에 대해서도 공부를 해보시면 좋을 것 같네요! 
[notification과 delegate를 비교하는 글입니다.](https://wlgusdn700.tistory.com/45)

추가로 해주신 질문에 대해서는 다른 방법은 저도 잘 모르겠네요. 🥲  저도 찾아보고 알게되면 공유해드리겠습니다!

2. 일단 아래 질문에 대해서는 
> 이전 방식에서 Navigation Bar의 Title 과 뒤로가기 버튼이 자동적으로 생성~

navigation viewcontroller에 대해서 공부를 해보시면 좋을 것 같네요. 🙏🏻
navigation viewcontroller에서는 navigationBar를 보여주고 말지 결정할 수 있습니다.

> 이에, 스토리보드를 보면 FruitStockViewController에 navigation title과 item들이 정상적으로 표시되는데도 불구하고 네비게이션 컨트롤러가 아닌 VC로 화면 전환을 바로하면 왜 네비게이션 바가 노출되지 않는지 이유가 궁금합니다.

또한 navigation viewcontroller에서 **present와 push**에 대해서도 같이 알아보시면 질문에 대한 답변이 될 수 있을 것 같네요. 🙏🏻
지금의 구현하신 방법으로도 문제가 해결되긴 했지만 navigationViewController 에서 또 다른 navigationViewController를 present 하기 때문에 의도하신 것을 해결은 했지만 좋은 방법은 아닙니다. 

추가로 Modal의 경우에는 navigationBar가 기본적으로 생성되지 않기 때문에 직접 navigationBarView를 생성하고, button을 달아서 동작하게 할 수 있습니다.
이 부분에 대해서는 modal에 대해서 추가적으로 알아보시면 도움이 될 것 같네요.

2-2. 화면 간 데이터를 전송하는 방법에 대해서는 아래 리뷰에서 코멘트로 남겨두었습니다.

3. 수꿍과 바드가 알아보신 것처럼 navigationBar의 height를 변경할 수는 없습니다. modal 형태라면 위에서 말했듯이 navigationBar 처럼 생긴 뷰를 만들기 때문에 height를 조절할 수 있을 것이고, 그렇지 않은 경우라면 title의 fontsize를 변경하면 사이즈가 커질 것 같은데, 이 부분은 지금 꼭 확인하지 않아도 될 것 같네요 🙂

4. stepper의 경우에는 기획서에 있는 것이 Xcode 11이후부터 새로운 디자인으로 변경되었습니다. 커스텀을 하는 방법도 있지만 이번 프로젝트에서 의도한 내용은 아닌 것 같으니 현재 디자인으로 하셔도 될 것 같습니다. 👍🏻

bar-d committed:
style: 함수 네이밍 변경 및 함수 은닉화

Jeon-Minsu committed:
refactor:  UIButton extension 추가로 인한 기존 함수 코드 중복 회피 및 네이밍 수정

Jeon-Minsu committed:
refactor: FruitStockViewController 내 타입 프로퍼티를 인스턴스 프로퍼티로 변경 및 이에 따른 함수 내 코드 수정

Jeon-Minsu commented:
안녕하세요 제이크!! @jryoun1 

전체 코멘트와 세부 코멘트(1개 제외)에 대하여 제이크의 피드백을 반영하여 코드를 수정해보았습니다!

하지만, 1개의 세부 코멘트에 관하여 난황을 겪고 있어 따로 DM 드립니다!

Jeon-Minsu committed:
refactor: 코드 중복을 회피하기 위하여 각각의 Label이 아닌 Fruitstore Class 인스턴스 전달

bar-d committed:
refactor:  레이블 및 스테퍼 콜렉션 타입으로 변경 후 값들을 변경해주는 함수를 for문으로 작성

bar-d committed:
refactor: fruitWarehouse 싱글톤 적용

Jeon-Minsu commented:
제이크의 피드백 덕분에 세부 코멘트 하나까지 수정해볼 수 있었습니다!!
감사합니다 제이크!!! 이에 제이크의 모든 피드백에 대하여 의견 남겨보았습니다!!
이번에도 잘 부탁드려요!! 👍 👍

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,65 @@
+
+class FruitStockViewController: UIViewController {
+    @IBOutlet private var fruitsLabel: [UILabel]!
+    @IBOutlet private var fruitsStepper: [UIStepper]!
```
배열을 사용해서 코드의 중복을 줄여주셨네요! 👏

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+    
+    private func updateStepperValue() {
+        for fruits in 0..<FruitStore.fruitWarehouse.count {
```
index를 의미하는 것 같은데 fruits란 네이밍은 오해를 일으킬 것 같아요.

Jeon-Minsu replied:
좀 더 확실하게 구분할 수 있도록 네이밍을 `index`로 변경해보았습니다!

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -90,10 +93,56 @@ class JuiceOrderViewController: UIViewController {
+}
+
+extension UIButton {
+    func setUpAttributes() {
```
UIButton을 extension해서 함수를 잘 만들어주셨네요 👍🏻
lineBreakMode, textAlignment, font도 매개변수로 받아와서 쓸 수 있다면 다양하게 사용할 수 있을 것 같아요!

bar-d replied:
저희도 그렇게 생각을 했는데 현재 사항에서는 똑같은 폰트와 사이즈등 동일하기에 매개변수 없이 함수를 만들어주었습니다!  
다음에는 매개변수를 넣어 더 활용해보겠습니다

bar-d replied:
저희도 그렇게 생각을 했는데 현재 사항에서는 똑같은 폰트와 사이즈등 동일하기에 매개변수 없이 함수를 만들어주었습니다!  
다음에는 매개변수를 넣어 더 활용해보겠습니다

Jeon-Minsu replied:
기존의 `setUp()` 함수에 파라미터를 추가하여 `setUp(lineBreakMode: NSLineBreakMode, textAlignment: NSTextAlignment, font: UIFont)` 와 같이 수정하여 좀 더 제작자의 의도에 맞게 변경할 수 있도록 확장성 높은 코드로 변경해보았습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -90,10 +93,56 @@ class JuiceOrderViewController: UIViewController {
+        present(modalViewController, animated: true)
+    }
+    
+    private func sendFruitLabelText(_ controller: FruitStockViewController) {
```
함수명이 명확하지 않은 것 같습니다!

Jeon-Minsu replied:
이전에 labelText를 보내던 함수여서 위와 같은 네이밍이 남아있었네요!
현재는 FruitStore를 보내주는 함수이므로, `sendFruitStoreData(to controller: FruitStockViewController)`로 변경해보았습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -90,10 +93,56 @@ class JuiceOrderViewController: UIViewController {
+    }
+    
+    private func setUpButton() {
+        strawberryBananaJuiceButton.setUpAttributes()
```
훨씬 깔끔해졌네요 💯
다만 네이밍이 strawberryBananaJuiceButton.setup(~)이런식이면 더 읽기 좋을 것 같아요.

Jeon-Minsu replied:
기존의 `setUp()` 함수에 파라미터를 추가하여 `setUp(lineBreakMode: NSLineBreakMode, textAlignment: NSTextAlignment, font: UIFont)` 와 같이 수정하여 좀 더 제작자의 의도에 맞게 변경할 수 있도록 확장성 높은 코드로 변경해보았습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -5,37 +5,37 @@
 
 class FruitStore {
-     private(set) var fruitWarehouse: Dictionary<Fruit, Int> = [
+    static var fruitWarehouse: Dictionary<Fruit, Int> = [
```
static이 된 이유가 있을까요??

Jeon-Minsu replied:
두 개의 뷰컨트롤러에서 접근을 하고 있어 싱글톤으로 생성하여 공동의 유일한 인스턴스로 사용해보는 것이 더욱 좋지 않을까라고 생각하여 다음과 같이 바꿔보았습니다. 하지만 과일 재고를 재고해보니, 굳이 의존성 주입을 통하여 `fruitStore` 데이터를 전달하였기 때문에, `fruitWarehouse`를 싱글톤이 아닌 하나의 인스턴스로 호출하여도 충분하다는 생각이 들어 원래의 `private(set) var`로 수정하였습니다.

jryoun1 replied:
싱글톤으로 사용을 해야할 이유가 있다면 싱글톤으로 객체를 생성하고 사용하면 됩니다. 
다만 개인적으로 vc간의 데이터를 전송하기 위해서 fruitStore를 싱글톤으로 만드는 것은 바람직하지 않다고 느껴져서 말씀드렸습니다.

추가로 싱글톤을 사용하기 위해서는 외부에서 클래스를 생성하지 못하게 클래스 생성자의 접근제어를 private으로 막아주어야 싱글톤이라고 할 수 있습니다. 이 부분에 대해서는 추후에 싱글톤을 공부해보시면서 확인해보시면 좋겠네요!👍🏻

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,65 @@
+    @IBOutlet private var fruitsLabel: [UILabel]!
+    @IBOutlet private var fruitsStepper: [UIStepper]!
+    
+    var fruitStore: FruitStore = FruitStore()
```
여기서 fruiteStore 객체를 생성하는 이유가 있을까요? 👀

또한 juiceMaker가 아닌 fruiteStore 를 넘겨받는 이유가 궁금합니다!

Jeon-Minsu replied:
의존성 주입을 통해 `JuiceOrderViewController`의 `FruitStore` 데이터를 전달하고자 `FruitStockViewController`에 이를 받을 프로퍼티를 생성해줄 필요가 있었습니다. 초기화를 위하여 `FruitStore()`을 생성하였는데 좀 더 의도를 명확히 밝히기 위하여 `var fruitStore: FruitStore? = nil`와 같이 빈 값으로 먼저 설정해보는 식으로 수정해보았습니다.

jryoun1 replied:
```swift
var fruitStore: FruitStore? 
```
알고 계시겠지만 옵셔널인 경우에는 반드시 초기값이 필요한 것은 아니라서 위와 같이 작성해도 됩니다! 
초기값이 nil이어야만 한다면 nil을 넣어주는 것이 맞구요 🙂

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,65 @@
+    
+    @IBAction private func changeValueOfStrawberryStepper(_ sender: UIStepper) {
+        updateFruitLabel()
+        postUpdateFruitStockNotification()
```
이 함수는 stepper의 값이 변경될 때마다 매번 호출되어야할까요??

Jeon-Minsu replied:
`FruitStockViewController`의 화면이 사라져 `JuiceOrderViewController`로 돌아올 때에만 알림을 송신해주면 되므로, `postUpdateFruitStockNotification()`를 `closeButtonTouchedUpInside()` 함수 내로 위치를 변경해주었습니다. 그리고 기존의 `updateFruitLabel()` 함수는 `Stepper`를 누를때마다 재고 레이블을 최신화해주어야 하기에 매번 호출되어야 한다고 생각합니다.

jryoun1 replied:
넵 제가 말씀드린 부분은 39번라인의 postUpdateFruitStockNotification() 였습니다!
updateFruitLabel()은 매번 호출되어야할 것 같네요!

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+    
+    private func postUpdateFruitStockNotification() {
+        center.post(name: .updateFruitStock, object: nil, userInfo: ["딸기재고": fruitsLabel[0].text ?? "", "바나나재고": fruitsLabel[1].text ?? "", "파인애플재고": fruitsLabel[2].text ?? "", "키위재고": fruitsLabel[3].text ?? "", "망고재고": fruitsLabel[4].text ?? ""])
```
post를 할 때 userInfo를 사용해보셨네요 👍🏻
이렇게 다시 값들을 넘기면 결국 처음에 있던 코드와 큰 차이가 없을 것 같습니다.
똑같이 100개 생긴다면 100개를 적어서 보내주어야할 것 같네요.

질문이 있습니다 🙋‍♂️
1. 클래스와 구조체의 차이는 무엇일까요?
2. fruitStore는 왜 클래스로 선언이 되었을까요?

Jeon-Minsu replied:
위와 같이 하나하나 코딩을 하기 보다는 `center.post(name: .updateFruitStock, object: nil, userInfo: ["updatedFruitStock": fruitsLabel ?? []])`와 같이 UILabel 배열을 송신하여 수신을 받는 뷰컨트롤러에서 반복문을 통해 각각의 과일 재고를 최신화하는 방법으로 바꾸어보았습니다!

Jeon-Minsu replied:
1. 클래스와 구조체의 차이는 무엇일까요?

두 타입의 가장 명확한 차이는 클래스는 참조타입이고, 구조체는 값 타입이라는 점입니다. 이에 한 구조체에서 다른 구조체로의 할당은 여전히 동일한 메모리 주소에 두 인스턴스를 모두 저장합니다. 하지만 하나의 구조체가 수정되면 `Swift` 메모리 관리는 이를 별도의 메모리 위치에 저장하기에 단순히 값을 복사하여 첫 번째 구조체의 속성 수정은 다른 구조체에 영향을 미치지 않습니다.
반면, 클래스의 경우 한 클래스에서 다른 클래스로의 할당은 두 클래스 인스턴스가 이제 동일한 메모리 주소를 가리키는 방식으로 참조를 복사합니다. 따라서 한 클래스 인스턴스가 수정되면 다른 인스턴스도 영향을 받습니다. 

저장 방식에도 차이가 있습니다. 구조체는 인스턴스 데이터를 모두 스택에 저장하는 반면, 클래스는 인스턴스 데이터를 힙에 저장하고, 해당 힙을 가리키는 변수를 스택에 저장합니다. 그리고 상속의 경우, 클래스는 단일 상속이 가능하나, 구조체는 상속이 불가능합니다.

2. `fruitStore`는 왜 클래스로 선언이 되었을까요?

`fruitStore`는 값 타입인 구조체를 사용하기 보다는 참조 타입인 클래스를 사용하는 편이 더욱 절절하다고 생각합니다. `Struct` 내 `Class`를 생성하면 `Class` 내부에서 재고의 변화는 참조타입이기 때문에 아무리 다수의 인스턴스로 생성되어도 모두 똑같은 값을 가집니다. 즉,  쥬스메이커가 몇 대가 있든 `FruitStore` 내 과일 재고의 양은 모두 공유하고 있음을 의미합니다. 쥬스메이커는 각각 하나의 쥬스 제조 기계로 서로 독립적이라고 보면 이해하기 편합니다. 예를 들자면, 쥬스 메이커가 여러 대이면 공동의 과일 재고를 가지고 있는 상태에서 어느 쥬스메이커는 3개의 쥬스를 만들고, 다른 쥬스메이커는 5대의 쥬스를 재고만 떨어지지 않는한 서로 관여없이 만들 수 있다는 의미에서 이해하였습니다.

또한, `FruitStore`은 현재 5개의 과일밖에 있지는 않지만, 사업이 번창하면 다양한 과일을 선보일 필요가 있다고 생각합니다. 현재는 하나의 소규모 과일 가게의 개념으로 `FruitStore`을 정의하였는데, 하나의 과일 대형마트, 심지어 과일 백화점으로 그 규모가 커질 수 있습니다. 따라서 이를 구조체로 정의하기 보다는 클래스로 정의하여 상속을 받는 편이 더욱 확장성있는 방법이라고 판단하였습니다.

jryoun1 replied:
2번에서 말씀하신 부분 중에서 제가 잘못이해를 한 것인지 모르겠지만 헷갈리는 부분이 있어서 남겨봅니다.
>  Struct 내 Class를 생성하면 Class 내부에서 재고의 변화는 참조타입이기 때문에 아무리 다수의 인스턴스로 생성되어도 모두 똑같은 값을 가집니다. 즉, 쥬스메이커가 몇 대가 있든 FruitStore 내 과일 재고의 양은 모두 공유하고 있음을 의미합니다.

```swift
let jM1 = JuiceMaker()
let jM2 = JuiceMaker()
try? jM1.fruitStore.changeStock(fruit: .strawberry, amount: 20)
print(jM1.fruitStore.fruitWarehouse[.strawberry])
print(jM2.fruitStore.fruitWarehouse[.strawberry])
```
jM1 객체 내부의 fruitStore 인스턴스와 jM2 객체 내부의 fruitStore 인스턴스는 서로 다른 인스턴스 입니다. 
위의 예시 코드를 확인해보시면 서로 다른 값이 나오게 될 것 입니다. 
만약 두 분이 말씀하신대로 모든 쥬스메이커가 같은 fruitStore를 공유하게 하고 싶다면 fruitStore가 싱글톤이어야만 가능할 것 입니다.

또한 수정한 부분에 대해서 말씀드리자면
> 위와 같이 하나하나 코딩을 하기 보다는 center.post(name: .updateFruitStock, object: nil, userInfo: ["updatedFruitStock": fruitsLabel ?? []])와 같이 UILabel 배열을 송신하여 수신을 받는 뷰컨트롤러에서 반복문을 통해 각각의 과일 재고를 최신화하는 방법으로 바꾸어보았습니다!

이러한 방법도 있겠지만, 제가 의도했던 부분은 JuiceOrderViewController에서 생성한 fruitStore 인스턴스를 FruitStockViewController로 전달해주었기 때문에, FruitStockViewController에서 fruitStore의 fruitWarehouse 값들을 변경하게 되면 JuiceOrderViewController로 돌아왔을 때 fruitStore 인스턴스의 fruitWarehouse 값을 확인하면 변경된 값을 확인할 수 있는 방법을 말씀드린 것입니다.
즉, 요약하자면 아래와 같은 과정입니다. 
1. JuiceOrderViewController에서 생성한 juiceMaker 객체 내부의 fruitStore 인스턴스 FruitStockViewController로 전달
2. FruitStockViewController에서는 전달받은 fruitStore 인스턴의 fruitWarehouse값들을 수정
3. FruitStockViewController에서 닫기를 눌러서 다시 JuiceOrderViewController로 이동
4. JuiceOrderViewController에서 fruitStore 인스턴스의 값들을 확인해서 label들을 업데이트

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -6,30 +6,34 @@
-    @IBOutlet private weak var pineappleLabel: UILabel?
-    @IBOutlet private weak var kiwiLabel: UILabel?
-    @IBOutlet private weak var mangoLabel: UILabel?
+    let juiceMaker = JuiceMaker()
```
이 juiceMaker도 밖에서 사용될 일이 있을까요?

Jeon-Minsu replied:
은닉화 작업 완료하였습니다!

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -90,10 +93,56 @@ class JuiceOrderViewController: UIViewController {
+        
     private func goToFruitStockViewController() {
-        guard let controller = self.storyboard?.instantiateViewController(withIdentifier: "FruitStockViewController") else { return }
+        guard let controller = self.storyboard?.instantiateViewController(withIdentifier: "FruitStockViewController") as? FruitStockViewController else { return }
```
타입캐스팅을 추가해주셨네요! 기존과 차이가 무엇일까요? 👀

Jeon-Minsu replied:
단순 화면 전환은 다운 캐스팅 없이도 충분히 가능하지만  `FruitStockViewController`에서 정의한 프로퍼티인 `fruitStore`를 사용하기 때문에 다운 캐스팅이 필요합니다. 다운 캐스팅을 하지 않으면, 다음의 에러가 나타납니다.

"Cannot convert value of type 'UIViewController' to expected argument type 'FruitStockViewController'"

이는 `FruitStockViewController`가 아닌 `UIViewController` 타입의 인스턴스가 생성되기 때문입니다.

Jeon-Minsu committed:
refactor: 함수 내부 코드 수정을 통한 중복 제거 및 네이밍 변경
- postUpdateFruitStockNotification() 함수 내 userInfo 데이터를 개별 데이터에서 UILabel 배열로 변경
- changeValueOfStrawberryStepper() 함수 내 postUpdateFruitStockNotification() 함수 제거
- closeButtonTouchedUpInside 함수 내 postUpdateFruitStockNotification() 함수 추가
- fruitWarehouse 접근 수준을 static에서 private(set)으로 변경
- juiceMaker 인스턴스 은닉화 설정
- 함수 내 for문의 fruits를 index로 네이밍 변경
- sendFruitLabelText()함수의 네이밍을 sendFruitStoreData(to:)으로 변경 및 아규먼트 레이블 추가
- tryUpdateFruitStock() 함수 내부 중복 제거
- setUpButton함수의 네이밍을 setUp(lineBreakMode:textAlignment:font:)으로 변경 및 파라미터 추가

Jeon-Minsu commented:
제이크의 피드백을 반영하여 코드를 수정해보았습니다!! @jryoun1

bar-d committed:
docs: README.md 파일 수정

bar-d committed:
Merge branch 'ic_6_sookoong' of https://github.com/Jeon-Minsu/ios-juice-maker into ic_6_sookoong

Jeon-Minsu committed:
refactor: 전역변수 center 삭제

jryoun1 commented:
@Jeon-Minsu @bar-d 
수정하신 부분에 대해서는 추가로 코멘트 남겨드렸습니다. 👏

[해당 부분](https://github.com/yagom-academy/ios-juice-maker/pull/237#discussion_r872197178)은 이번 프로젝트에서 수정해보셔도 되고, 다음 프로젝트에서 클래스를 사용할 때 적용해보셔도 됩니다. 🙏🏻
step3는 여기서 머지하도록 할게요~ 
3주 동안 고생하셨습니다!
